File: Python/Easy/1920_build_array_from_permutation.py
"""
1920. Build Array from Permutation

Given a zero-based permutation nums (0-indexed), build an array ans of the
same length where ans[i] = nums[nums[i]] for each 0 <= i < nums.length and
return it.

A zero-based permutation nums is an array of distinct integers from
0 to nums.length - 1 (inclusive).

Problem Link: https://leetcode.com/problems/build-array-from-permutation/

Examples:
    Example 1:
        Input: nums = [0,2,1,5,3,4]
        Output: [0,1,2,4,5,3]

    Example 2:
        Input: nums = [5,0,1,2,3,4]
        Output: [4,5,0,1,2,3]

Constraints:
    1 <= nums.length <= 1000
    0 <= nums[i] < nums.length
    The elements in nums are distinct.
"""

from typing import List


class Solution:
    def buildArray(self, nums: List[int]) -> List[int]:
        """
        Build an array `ans` where `ans[i] = nums[nums[i]]` using a new list
        for storing the result.

        Time Complexity: O(n), where n is the length of the input list nums.
        Space Complexity: O(n) due to the extra space used by the `ans` list.

        Args:
            nums (List[int]): The input list of integers representing a
            zero-based permutation.

        Returns:
            List[int]: The list generated from the permutation.
        """
        n: int = len(nums)
        ans: List[int] = [0] * n

        for i in range(n):
            ans[i] = nums[nums[i]]

        return ans

    def buildArrayInPlace(self, nums: List[int]) -> List[int]:
        """
        Build an array `ans` where `ans[i] = nums[nums[i]]` without using
        extra space. This method leverages the properties of modular arithmetic
        to store both the old and new values in the same array.

        Time Complexity: O(n), where n is the length of the input list nums.
        Space Complexity: O(1), since no extra space is used except for the
        input list.

        Args:
            nums (List[int]): The input list of integers representing a
            zero-based permutation.

        Returns:
            List[int]: The list generated from the permutation.
        """
        n: int = len(nums)

        for i in range(n):
            nums[i] = nums[i] + (nums[nums[i]] % n) * n

        for i in range(n):
            nums[i] = nums[i] // n

        return nums


if __name__ == "__main__":
    solution = Solution()

    # Test cases
    test_cases = [
        ([0, 2, 1, 5, 3, 4], [0, 1, 2, 4, 5, 3]),
        ([5, 0, 1, 2, 3, 4], [4, 5, 0, 1, 2, 3]),
        ([0], [0]),
        ([4, 3, 2, 1, 0], [0, 1, 2, 3, 4])
    ]

    for numbers, expected in test_cases:
        # Test the basic method
        result = solution.buildArray(numbers.copy())
        assert result == expected, f"""
            Test case failed: expected {expected}, got {result}
            """

        # Test the in-place method
        result_in_place = solution.buildArrayInPlace(numbers.copy())
        assert result_in_place == expected, f"""
            Test case failed: expected {expected}, got {result_in_place}
            """

    print("All test cases passed!")

File: Golang/Easy/1920_build_array_from_permutation.go
package main

import "fmt"

// buildArray returns ans where ans[i] = nums[nums[i]] using an O(n) auxiliary
// array.
//
// Time complexity:  O(n)
// Space complexity: O(n) — for the output array.
func buildArray(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	for i := 0; i < n; i++ {
		ans[i] = nums[nums[i]]
	}
	return ans
}

// buildArrayInPlace solves the follow-up: produce nums[nums[i]] in-place with
// O(1) extra space.
//
// Trick: pack new and old values into each slot. Since nums[i] < 1000 by the
// problem constraint, encode as old + 1000 * (new % 1000) — the high digits
// hold the answer, the low digits preserve the original so later indices can
// still recover it. A second pass divides out the encoding.
//
// Time complexity:  O(n)
// Space complexity: O(1)
func buildArrayInPlace(nums []int) []int {
	n := len(nums)
	for i := 0; i < n; i++ {
		nums[i] += 1000 * (nums[nums[i]] % 1000)
	}

	for i := 0; i < n; i++ {
		nums[i] /= 1000
	}
	return nums
}

func main() {
	testCases := []struct {
		nums     []int
		expected []int
	}{
		{[]int{0, 2, 1, 5, 3, 4}, []int{0, 1, 2, 4, 5, 3}},
		{[]int{5, 0, 1, 2, 3, 4}, []int{4, 5, 0, 1, 2, 3}},
		{[]int{0}, []int{0}},
		{[]int{4, 3, 2, 1, 0}, []int{0, 1, 2, 3, 4}},
	}

	funcs := []struct {
		label string
		fn    func([]int) []int
	}{
		{"buildArray       ", buildArray},
		{"buildArrayInPlace", buildArrayInPlace},
	}

	for _, f := range funcs {
		for i, tc := range testCases {
			input := append([]int(nil), tc.nums...)
			result := f.fn(input)
			status := "PASS"
			if fmt.Sprintf("%v", result) != fmt.Sprintf("%v", tc.expected) {
				status = "FAIL"
			}
			fmt.Printf("%s | Example %d | Input: %v | Expected: %v | Result: %v | [%s]\n",
				f.label, i+1, tc.nums, tc.expected, result, status)
		}
	}
}
